### ISSUE TITLE
GSoC Project Introduction

### ISSUE THREAD

**Issue by vikasTmz** [lagadic/visp_cao_editor#1]
Hi,
My name is **Vikas Thamizharasan** and I have been selected to work with ViSP on this project. 
Here is a link to my [Proposal](https://docs.google.com/document/d/1duq67mYs2-62eQLmg_OPNgahRIaNFDyB8TkO6fpkOrQ/edit). Feel free to suggest any improvements before 30<sup>th</sup> May, which is when the first coding period commences. 

My first two weeks would consist of working on **Phase 1, _Task 1.1_** (covered in page 4-8).
[Page 10](https://docs.google.com/document/d/1duq67mYs2-62eQLmg_OPNgahRIaNFDyB8TkO6fpkOrQ/edit#bookmark=id.9c3qu21gch5c) contains potentially useful links to creating the blender plugin.

This is an introductory thread. I plan to create new issues for each task and reference them in their respective PR. Any suggestions on how I should proceed on working on a new project, documentation and coding styles will be extremely helpful.

**Comment by fspindle** [comment]
Ok

We stated a coding style [page](https://github.com/lagadic/visp/wiki/coding-style-guideline) that may evolve. If info is missing don't hesitate to tell us

**Comment by s-trinh** [comment]
I think we should use two subdirectories in `visp_cao_editor`:

- one for the Blender plugin

- one for the Qt Creator application

As usual, only needed files should be versioned + some `README` file for quick instructions.

We should add / recall the features to be done in this thread. For instance for the Blender part:

- create a Python Blender plugin to export a regular CAD model to a `.cao` model file
  - [ ] direct export to `.cao` model file, with some UI for non trivial primitives (circle, cylinder)
  - [ ] add possibility to export / create a hierarchical `.cao` model file, use the object grouping in Blender?
  - [ ] create a mini tutorial about how to simplify a complex 3D model in Blender?
- create a Python Blender plugin to display a custom UI
  - [ ] dialog box to set to a selected object a name and / or get the name from the object property
  - [ ] dialog box to set LOD parameter to a selected object
- create a Python Blender plugin to import a `.cao` model file
  - [ ] read lines geometric primitives + display in Blender
  - [ ] read face geometric primitives + display in Blender
  - [ ] read circle geometric primitives + display in Blender
  - [ ] read cylinder geometric primitives + display in Blender
  - [ ] parse the optional settings (name, LOD parameters) in the `.cao` file to be able to display this information in Blender, via a contextual menu
- quick demo / validation
  - [ ] export `CubeSat` model after simplify it with Blender in `.cao`
  - [ ] set some parameters to the model + export in `.cao`
  - [ ] read the simplified `CubeSat` in `.cao` in Blender with the corresponding optional settings if any

This is just an example, it can be modified / completed.

---

Similarly to the other GSoC project, it could be a good idea to fill the project page (one example [here](https://github.com/lagadic/visp_unity/projects/1) for the ViSP Unity repository):

- split the first and second part of the project as some tasks to do
- create for each task a card or an issue from the project [page](https://github.com/lagadic/visp_cao_editor/projects/1)?
- a task could be one bullet in this example?

**Comment by vikasTmz** [comment]
I've explored a bit into model simplification in Blender and most techniques help reduce polycount more than converting into real object visible contours. 

Among the techniques mentioned in **Task 1.3** in my proposal, Limited dissolve and Decimate Modifier are others that would help to reduce polycount. But the simplification done on the `cubesat` dataset (_from `cubesat1b.obj` to `cubsat1b_simple.obj`_) cannot be automated with the above techniques. Decimate modifier will alter the contours of the model. 

So with respect to the `cubesat` model simplification, are we looking to automate this? Or is a manual process (_selection+deletion+dissolve_) enough? By manual I mean it would take the user some understanding of blender options and tools to achieve the simplification.

**Comment by vikasTmz** [comment]
Also, the simplification should include removal of any texture mapping/UV mapping and Shading, right?

Since the `.cao` format does not store any of these information unlike other 3D formats, if these above properties of the 3D model **are not needed to be removed**, then on exporting should these be stored in a separate file (eg: the `.mtl` file created when exported to `.obj`)? So on importing the `.cao` model, these properties can be rendered in blender.

Also does the orientation of the 3D model in blender with respect to the world coordinate system 
matter when used in MBT?

**Comment by s-trinh** [comment]
> So with respect to the cubesat model simplification, are we looking to automate this? Or is a manual process (selection+deletion+dissolve) enough? By manual I mean it would take the user some understanding of blender options and tools to achieve the simplification.

I know just the basics in Blender. If it is not possible to (easily) automate the simplification process with Blender, my suggestion is to write a tutorial (+ maybe a video tutorial) about how to use the Blender tools to simplify a CAD model.

> Since the .cao format does not store any of these information unlike other 3D formats, if these above properties of the 3D model are not needed to be removed, then on exporting should these be stored in a separate file (eg: the .mtl file created when exported to .obj)? So on importing the .cao model, these properties can be rendered in blender.

These properties (texture mapping/UV mapping and Shading) are not needed and can be removed. It is an optional feature (only when all the required features are done and work) if you want to embed these properties in an additional file for nice rendering in Blender.

> Also does the orientation of the 3D model in blender with respect to the world coordinate system
> matter when used in MBT?

No we don't need this. We only need for the .cao CAD model to contain a list of points (+ some additional primitives, i.e. lines, faces, circles, cylinders, constructed from some specific 3D object points) that describe the model in the object frame.

**Comment by s-trinh** [comment]
@vikasTmz Hello Vikaz. Thank you for the first pull request.

We have discussed internally and we have summarized in the [project](https://github.com/lagadic/visp_cao_editor/projects/1) dashboard all the features we would like to have.

Feel free to comment about it. If something is not feasible easily, feel free to propose an alternative. The idea is to have a Blender tool that can allow to export/model easily a complex model to `.cao` model file.

Some optional or advanced features can be dropped (i.e. automatic simplification of a complex model in Blender).
